File: solutions/leetcode/2799.统计完全子数组的数目.py
#
# @lc app=leetcode.cn id=2799 lang=python3
#
# [2799] 统计完全子数组的数目
#

# @lc code=start
class Solution:
    def countCompleteSubarrays(self, nums: List[int]) -> int:
        n = len(set(nums))
        counter = defaultdict(int)

        lo, hi, ans = 0, 0, 0
        while hi < len(nums):
            counter[nums[hi]] += 1
            while len(counter) == n:
                ans += len(nums) - hi 
                counter[nums[lo]] -= 1
                if counter[nums[lo]] == 0:
                    del counter[nums[lo]]
                lo += 1
            hi += 1
        return ans


        
# @lc code=end

File: solutions/leetcode/2799.统计完全子数组的数目.cpp
/*
 * @lc app=leetcode.cn id=2799 lang=cpp
 *
 * [2799] 统计完全子数组的数目
 */

// @lc code=start
class Solution {
    public:
        int countCompleteSubarrays(vector<int>& nums) {
            int n = set<int>(nums.begin(), nums.end()).size();
    
            int lo = 0, hi = 0, ans = 0;
            map<int, int> mp;
            while (hi < nums.size()) {
                mp[nums[hi]]++;
                while (mp.size() == n) {
                    ans += nums.size() - hi;
                    mp[nums[lo]]--;
                    if (mp[nums[lo]] <= 0) 
                        mp.erase(nums[lo]);
                    lo++;
                }
                hi++;
            }
            return ans;
        }
    };
// @lc code=end

